### PR TITLE
Added a customQualitySelector config option

### PR DESCRIPTION
**Pull request by bfitec**
### Link to related issue (if applicable)

This address issues #1455 #1372 #1118 #1446 #1372 #652.

### Summary of proposed changes

Added a customQualitySelector config option to allow for a custom quality change function overiding the default one.  After adding this option creating auto resolution selections with hls.js will be trivial. Note: When I built this I had TypeErrors in plugin 'gulp-imagemin' - not sure if this is expected behavior.

### Checklist
- [x] Use `develop` as the base branch
- [x] Exclude the gulp build (`/dist` changes) from the PR
- [x] Test on [supported browsers](https://github.com/sampotts/plyr#browser-support)

### CODE REVIEW

**Comment by bfitec** [comment]
@sampotts 
Any possibility of getting this added? I'm currently running a custom build and would love to be able to get back on this repo : ) THANK YOU for all the work you all do on this project!

**Comment by bfitec** [comment]
@sampotts bump

**Comment by bfitec** [comment]
@sampotts I would really really like to see this merged, what can I do to help? We'd really like to increase adoption of Plyr to improve the stability of our Plyr (which we're using in production).

**Comment by Pandakls** [comment]
You may wan't remove the console.log for a final merge.
I looked at it but i'm not competent enough to approve your modification.

Can someone merge this ? It would be really great to handle basic HLS feature as quality change.

**Comment by bfitec** [comment]
@Pandakls sorry didn't realize there was a console.log in the merge, my bad.

I don't think there's any use in fixing it as I've gotten no response from @sampotts. There are tons of PRs and issues and seemingly no movement. Right now we're running off a fork in production.

**Comment by efriandika** [comment]
@sampotts please

File: src/js/controls.js [hunk 1]
```diff
@@ -900,6 +900,7 @@ const controls = {
         if (is.array(options)) {
             this.options.quality = dedupe(options).filter(quality => this.config.quality.options.includes(quality));
         }
+        console.log('This.option:',this.options);
```
Comment (sampotts) [inline src/js/controls.js]: Can you remove this? If you need logging you can use `this.debug.log`

File: src/js/controls.js [hunk 2]
```diff
@@ -1601,6 +1602,7 @@ const controls = {
 
         // Set available quality levels
         if (this.isHTML5) {
+            console.log('Result:',html5.getQualityOptions.call(this));
```
Comment (sampotts) [inline src/js/controls.js]: As above re: logging

File: src/js/html5.js [hunk 3]
```diff
@@ -30,6 +30,12 @@ const html5 = {
 
     // Get quality levels
     getQualityOptions() {
+        const player = this;
+        console.log('GET QUAL:',player);
+        if (player.config.hasOwnProperty('forceQualityOptions')){
+          return player.config.forceQualityOptions;
```
Comment (sampotts) [inline src/js/html5.js]: Please use 4 space tabs

File: src/js/html5.js [hunk 4]
```diff
@@ -30,6 +30,12 @@ const html5 = {
 
     // Get quality levels
     getQualityOptions() {
+        const player = this;
+        console.log('GET QUAL:',player);
```
Comment (sampotts) [inline src/js/html5.js]: As above re: logging

File: src/js/html5.js [hunk 5]
```diff
@@ -60,6 +66,17 @@ const html5 = {
                 return source && Number(source.getAttribute('size'));
             },
             set(input) {
+              // Use custom quality selector if set in config and return.
+              if (player.config.hasOwnProperty('customQualitySelector') && typeof(player.config.customQualitySelector) === 'function'){
```
Comment (sampotts) [inline src/js/html5.js]: Please use built-in type checking library - i.e. `is.function(...)` and `Object.keys(player.config).includes()`

**Comment by sampotts** [comment]
I'll manually copy the logic from this and clean it up. I realize it's been around for ages and I'm really sorry it's taken me so long to get back to it.

**Comment by bfitec** [comment]
> I'll manually copy the logic from this and clean it up. I realize it's been around for ages and I'm really sorry it's taken me so long to get back to it.

I can clean it up if you'd like me to, I was just working on it. No worries, thank you for checking it out! I know you're very busy and want to make my commits easy for you to approve, sorry about the formatting and console logs. This is my first PR to a public project. @sampotts

**Comment by sampotts** [comment]
> > I'll manually copy the logic from this and clean it up. I realize it's been around for ages and I'm really sorry it's taken me so long to get back to it.
> 
> I can clean it up if you'd like me to, I was just working on it. No worries, thank you for checking it out! I know you're very busy and want to make my commits easy for you to approve, sorry about the formatting and console logs. This is my first PR to a public project. @sampotts

Thanks for your hard work on it! Ideally the PR should only include files from `/src` and not the files in `/dist` as it makes reviewing the PR harder. It looks like you may have edited the built files though? You should edit the relevant files in `src/` and then run gulp, check the output and then PR the `/src` files.

**Comment by bfitec** [comment]
@sampotts Ah yes, I vaguely remember editing the compiled files... Would you like me to create close the PR and open a new one with these issues fixed?

**Comment by sampotts** [comment]
From what I can see, the main changes are quite small. I've just copied them into my branch so I'll test quickly, push and you can check if I'm correct.

**Comment by sampotts** [comment]
Have a look at this and see if that's what you wanted:
https://github.com/sampotts/plyr/commit/6ffaef35cf667d0e2e14227882a1a8e329b2c2c2

**Pull request closed without merging**